cpp/src/arrow/dataset/filter.cc

```diff
@@ -758,43 +797,107 @@ Result<std::shared_ptr<DataType>> FieldExpression::Validate(const Schema& schema
+      return std::move(out);
+    }
 
-  void operator()(const NotExpression& expr) { VisitExpression(*expr.operand(), *this); }
+    Result<std::shared_ptr<Expression>> operator()(const NotExpression& expr) {
+      ARROW_ASSIGN_OR_RAISE(auto op, Validate(*expr.operand()));
 
-  void operator()(const AndExpression& expr) {
-    VisitExpression(*expr.left_operand(), *this);
-    VisitExpression(*expr.right_operand(), *this);
-  }
+      if (op.type->id() != Type::BOOL) {
+        op.expr = op.expr->CastTo(boolean()).Copy();
+      }
+      return not_(std::move(op.expr));
+    }
 
-  void operator()(const OrExpression& expr) {
-    VisitExpression(*expr.left_operand(), *this);
-    VisitExpression(*expr.right_operand(), *this);
-  }
+    Result<std::shared_ptr<Expression>> operator()(const AndExpression& expr) {
+      ARROW_ASSIGN_OR_RAISE(auto lhs, Validate(*expr.left_operand()));
+      ARROW_ASSIGN_OR_RAISE(auto rhs, Validate(*expr.right_operand()));
 
-  void operator()(const ComparisonExpression& expr) {
-    VisitExpression(*expr.left_operand(), *this);
-    VisitExpression(*expr.right_operand(), *this);
-  }
+      if (lhs.type->id() != Type::BOOL) {
+        lhs.expr = lhs.expr->CastTo(boolean()).Copy();
+      }
+      if (rhs.type->id() != Type::BOOL) {
+        rhs.expr = rhs.expr->CastTo(boolean()).Copy();
+      }
+      return and_(std::move(lhs.expr), std::move(rhs.expr));
+    }
 
-  template <typename E>
-  void operator()(const E& expr) const {}
+    Result<std::shared_ptr<Expression>> operator()(const OrExpression& expr) {
+      ARROW_ASSIGN_OR_RAISE(auto lhs, Validate(*expr.left_operand()));
+      ARROW_ASSIGN_OR_RAISE(auto rhs, Validate(*expr.right_operand()));
 
-  std::vector<std::string> fields;
-};
+      if (lhs.type->id() != Type::BOOL) {
+        lhs.expr = lhs.expr->CastTo(boolean()).Copy();
+      }
+      if (rhs.type->id() != Type::BOOL) {
+        rhs.expr = rhs.expr->CastTo(boolean()).Copy();
+      }
+      return or_(std::move(lhs.expr), std::move(rhs.expr));
+    }
+
+    Result<std::shared_ptr<Expression>> operator()(const ComparisonExpression& expr) {
+      ARROW_ASSIGN_OR_RAISE(auto lhs, Validate(*expr.left_operand()));
+      ARROW_ASSIGN_OR_RAISE(auto rhs, Validate(*expr.right_operand()));
+
+      if (lhs.type->Equals(rhs.type)) {
+        return expr.Copy();
+      }
+
+      if (lhs.expr->type() == ExpressionType::SCALAR) {
+        lhs.expr = lhs.expr->CastTo(rhs.type).Copy();
+      } else {
+        rhs.expr = rhs.expr->CastTo(lhs.type).Copy();
+      }
+      return std::make_shared<ComparisonExpression>(expr.op(), std::move(lhs.expr),
+                                                    std::move(rhs.expr));
+    }
+
+    Result<std::shared_ptr<Expression>> operator()(const Expression& expr) const {
+      return expr.Copy();
+    }
+
+    const Schema& schema_;
+  } visitor = {schema};
+
+  return VisitExpression(expr, visitor);
```

Comment:
I wonder if at some point this functionality should be implemented as a virtual function on each expression type, rather than as a visitor.
(is it planned to make it possible to define custom expressions?)

Reply:
Custom expressions are already available, and I can refactor this to a virtual method without much trouble.

Reply:
I think it should probably remain a non member, however. The intention of Expression is to provide a simple expression tree which others mutate/validate/... so I'd like to keep the interface as minimal as possible